streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -513,107 +514,166 @@ private long computeLatency() {
+                                      final long recordsProcessedBeforeCommit) {
+
+        long totalProcessedEachRound;
+        long totalProcessedSinceLastCommit = 0;
+        // Round-robin scheduling by taking one record from each task repeatedly
+        // until no task has any records left
+        do {
+            totalProcessedEachRound = 0;
+            for (StreamTask task : tasks.values()) {
+                int numBuffered = task.process();
+                // we processed one record,
+                // and more are buffered waiting for the next round
+                if (numBuffered > 0) {
+                    totalProcessedEachRound++;
+                    totalProcessedSinceLastCommit++;
                 }
             }
+            if (recordsProcessedBeforeCommit != UNLIMITED_RECORDS &&
+                totalProcessedSinceLastCommit >= recordsProcessedBeforeCommit) {
+                totalProcessedSinceLastCommit = 0;
+                long processLatency = computeLatency();
+                streamsMetrics.processTimeSensor.record(processLatency / (double) totalProcessedSinceLastCommit,
+                    timerStartedMs);
+                maybeCommit(this.timerStartedMs);
+            }
+        } while (totalProcessedEachRound != 0);
 
-            // try to process one fetch record from each task via the topology, and also trigger punctuate
-            // functions if necessary, which may result in more records going through the topology in this loop
-            if (totalNumBuffered > 0 || polledRecords) {
-                totalNumBuffered = 0;
-
-                if (!activeTasks.isEmpty()) {
-                    for (StreamTask task : activeTasks.values()) {
-
-                        totalNumBuffered += task.process();
+        // go over the tasks again to punctuate or commit
+        for (StreamTask task : tasks.values()) {
+            maybePunctuate(task);
+            if (task.commitNeeded())
+                commitOne(task);
+        }
 
-                        requiresPoll = requiresPoll || task.requiresPoll();
+        return totalProcessedSinceLastCommit;
+    }
 
-                        streamsMetrics.processTimeSensor.record(computeLatency(), timerStartedMs);
+    /**
+     * Adjust the number of records that should be processed by scheduler. This avoids
+     * scenarios where the processing time is higher than the commit time.
+     * @param prevRecordsProcessedBeforeCommit Previous number of records processed by scheduler.
+     * @param totalProcessed Total number of records processed in this last round.
+     * @param processLatency Total processing latency in ms processed in this last round.
+     * @param commitTime Desired commit time in ms.
+     * @return An adjusted number of records to be processed in the next round.
+     */
+    private long adjustRecordsProcessedBeforeCommit(final long prevRecordsProcessedBeforeCommit, final long totalProcessed,
```

Comment:
I'd only want to make this adjustment if processLatency > commitTime though. In most cases that would mean no adjustment is needed at all and all records that are polled are processed. E.g., with the default commit time of 30 seconds currently I don't want to make any adjustments at all.